libs/partners/ai21/langchain_ai21/chat_models.py

```diff
@@ -0,0 +1,168 @@
+
+    max_tokens: int = 16
+    """The maximum number of tokens to generate for each response."""
+
+    min_tokens: int = 0
+    """The minimum number of tokens to generate for each response."""
+
+    temperature: float = 0.7
+    """A value controlling the "creativity" of the model's responses."""
+
+    top_p: float = 1
+    """A value controlling the diversity of the model's responses."""
+
+    top_k_return: int = 0
+    """The number of top-scoring tokens to consider for each generation step."""
+
+    frequency_penalty: Optional[Penalty] = None
+    """A penalty applied to tokens that are frequently generated."""
+
+    presence_penalty: Optional[Penalty] = None
+    """ A penalty applied to tokens that are already present in the prompt."""
+
+    count_penalty: Optional[Penalty] = None
+    """A penalty applied to tokens based on their frequency in the generated responses."""
+
+    @property
+    def _llm_type(self) -> str:
+        """Return type of chat model."""
+        return "chat-ai21"
+
+    def _stream(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> Iterator[ChatGenerationChunk]:
+        raise NotImplementedError
+
+    async def _astream(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> AsyncIterator[ChatGenerationChunk]:
+        raise NotImplementedError
+
+    def _generate(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
+        system, ai21_messages = _convert_messages_to_ai21_messages(messages)
+
+        response = self.client.chat.create(
+            model=self.model,
+            messages=ai21_messages,
+            system=system or "",
+            num_results=self.num_results,
+            temperature=self.temperature,
+            max_tokens=self.max_tokens,
+            min_tokens=self.min_tokens,
+            top_p=self.top_p,
+            top_k_return=self.top_k_return,
+            stop_sequences=stop,
+            frequency_penalty=self.frequency_penalty,
+            presence_penalty=self.presence_penalty,
+            count_penalty=self.count_penalty,
+            **kwargs,
+        )
+
+        outputs = response.outputs
+        message = AIMessage(content=outputs[0].text)
+        return ChatResult(generations=[ChatGeneration(message=message)])
+
+    async def _agenerate(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
+        raise NotImplementedError
```

Comment:
Shouldn't it so the same as logic as agenerate of the LLM class?

Reply:
Done